### ISSUE TITLE
Network not converging after removing L2Normalization Layer

### ISSUE THREAD

**Issue by Prasad9** [zhreshold/mxnet-ssd#128]
I am training a VGG16_reduced network on SSD with all the default values and default network architecture provided in the repository. The learned weights of this model, I have to port to iOS for Apple CoreML through [MXNet-to-CoreMl](https://github.com/apache/incubator-mxnet/tree/master/tools/coreml) tool provided by MXNet. But for now, this tool only supports [12 layers](https://github.com/apache/incubator-mxnet/tree/master/tools/coreml#layers). 

In the default network provided for SSD, there is a L2Normalisation layer whose conversion is not supported currently. Hence, I removed that layer by setting value of -1 in [symbol/symbol_factory.py](https://github.com/zhreshold/mxnet-ssd/blob/master/symbol/symbol_factory.py#L39) file like this. 

```
ratios = [[1,2,.5], [1,2,.5,3,1./3], [1,2,.5,3,1./3], [1,2,.5,3,1./3], \
                [1,2,.5], [1,2,.5]]
normalizations = [-1, -1, -1, -1, -1, -1]   # Set first value to -1 from 20
steps = [] if data_shape != 300 else [x / 300.0 for x in [8, 16, 32, 64, 100, 300]]
```

However, removing this L2Normalization layer is leading to failure in convergence during training phase, i.e my cross-entropy and SmoothL1 values during training are always being shown as nan.
Training log:
```
$ python train.py 
Using mxnet as:
<module 'mxnet' from '/my/path/lib/python3.5/site-packages/mxnet/__init__.py'>
Warning: using pre-installed version of mxnet may cause unexpected error...
(export MXNET_EXAMPLE_SSD_DISABLE_PRE_INSTALLED=1) to prevent loading pre-installed mxnet.
[18:44:39] src/io/iter_image_det_recordio.cc:280: ImageDetRecordIOParser: /my/path/mxnet-ssd/data/train.rec, use 7 threads for decoding..
[18:44:49] src/io/iter_image_det_recordio.cc:333: ImageDetRecordIOParser: /my/path/mxnet-ssd/data/train.rec, label padding width: 350
[18:44:50] src/io/iter_image_det_recordio.cc:280: ImageDetRecordIOParser: /my/path/mxnet-ssd/data/val.rec, use 7 threads for decoding..
[18:44:51] src/io/iter_image_det_recordio.cc:333: ImageDetRecordIOParser: /my/path/mxnet-ssd/data/val.rec, label padding width: 350
INFO:root:Start training with (gpu(0)) from pretrained model /my/path/mxnet-ssd/model/vgg16_reduced
[18:44:53] src/nnvm/legacy_json_util.cc:190: Loading symbol saved by previous version v0.8.0. Attempting to upgrade...
[18:44:53] src/nnvm/legacy_json_util.cc:198: Symbol successfully upgraded!
INFO:root:Freezed parameters: [conv1_1_weight,conv1_1_bias,conv1_2_weight,conv1_2_bias,conv2_1_weight,conv2_1_bias,conv2_2_weight,conv2_2_bias]
[18:44:58] src/operator/././cudnn_algoreg-inl.h:112: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DE
FAULT to 0 to disable)
[18:45:04] src/operator/convolution.cu:119: This convolution is not supported by cudnn, MXNET convolution is applied.
INFO:root:Epoch[0] Batch [20]   Speed: 49.56 samples/sec        CrossEntropy=nan        SmoothL1=nan
INFO:root:Epoch[0] Batch [40]   Speed: 50.28 samples/sec        CrossEntropy=nan        SmoothL1=nan
INFO:root:Epoch[0] Batch [60]   Speed: 49.67 samples/sec        CrossEntropy=nan        SmoothL1=nan
INFO:root:Epoch[0] Batch [80]   Speed: 49.84 samples/sec        CrossEntropy=nan        SmoothL1=nan
INFO:root:Epoch[0] Batch [100]  Speed: 49.77 samples/sec        CrossEntropy=nan        SmoothL1=nan
```
 
In order to make my model converge, I added the batch normalization layer (as conversion tool supports this layer). I added it in [conv_act_layer function in symbol/common.py](https://github.com/zhreshold/mxnet-ssd/blob/master/symbol/common.py#L5) by setting the default value as True. This is my code line:
```
def conv_act_layer(from_layer, name, num_filter, kernel=(1,1), pad=(0,0), \
    stride=(1,1), act_type="relu", use_batchnorm=True):    #Use batch norm by default
```
Adding this batch norm layer is failing in convergence again. I even changed the learning rate to lower value and still the network is not converging.

Please can somebody tell me which layer should I add and where so as to make my network converge after I remove the L2Normalization layer.

Thanks in advance.

**Comment by zhreshold** [comment]
You can skip the l2norm layer and use a scalar to normalize it, say
```
# y = sym.L2Normalize(x)
y = x * 0.0001
```
If this trick don't work, you can choose other models, e.g. resent, which is better IMO.

**Comment by xioryu** [comment]
You need to use L2Norm with scale=20 such that the neural response of conv4_3 is not too high to explode. Even with bn, the numeric value of bn-scale is too large to learn (maybe 20x conv5?).
If you really want to remove L2Norm, you may try adding BN **every** conv-layer.
